src/codegen/browser/selectors.ts

```diff
@@ -0,0 +1,75 @@
+import { ElementSelector } from '@/schemas/recording'
+
+interface CssSelector {
+  type: 'css'
+  selector: string
+}
+
+interface GetByRoleSelector {
+  type: 'role'
+  role: string
+  name: string
+}
+
+interface GetByTestIdSelector {
+  type: 'test-id'
+  testId: string
+}
+
+export type NodeSelector = CssSelector | GetByRoleSelector | GetByTestIdSelector
+
+function getRoleSelector(selectors: ElementSelector): GetByRoleSelector | null {
+  if (selectors.role === undefined) {
+    return null
+  }
+
+  return {
+    type: 'role',
+    role: selectors.role.role,
+    name: selectors.role.name,
+  }
+}
+
+function getTestIdSelector(
+  selectors: ElementSelector
+): GetByTestIdSelector | null {
+  if (selectors.testId === undefined || selectors.testId.trim() === '') {
+    return null
+  }
+
+  return {
+    type: 'test-id',
+    testId: selectors.testId,
+  }
+}
+
+function getCssSelector(selectors: ElementSelector): CssSelector {
+  return {
+    type: 'css',
+    selector: selectors.css,
+  }
+}
+
+export function getNodeSelector(selector: ElementSelector): NodeSelector {
+  return (
+    getRoleSelector(selector) ??
+    getTestIdSelector(selector) ??
+    getCssSelector(selector)
+  )
```

Comment:
Selects the first non-null selector in order of execution. `getCssSelector` is guaranteed to always be a selector.